core/src/jsMain/kotlin/dev/fritz2/remote/http.kt

```diff
@@ -6,229 +6,306 @@ import org.w3c.fetch.*
 import org.w3c.files.Blob
 import org.w3c.xhr.FormData
 import kotlinx.browser.window as browserWindow
+import org.w3c.fetch.Response as FetchResponse
 
 
+/**
+ * creates a new [Request]
+ *
+ * @param baseUrl the common base of all urls that you want to call using the template
+ */
+fun http(baseUrl: String = "") = Request(url = baseUrl)
+
 /**
  * [Exception] type for handling http exceptions
  *
  * @property statusCode the http response status code
  * @property body the body of the error-response
  */
-class FetchException(val statusCode: Short, val body: String, val response: Response) : Exception(
+class FetchException(val statusCode: Int, val body: String, val response: Response) : Exception(
     "code=$statusCode, url=${response.url}, body=$body"
 )
 
 /**
- * creates a new [Request]
- *
- * @param baseUrl the common base of all urls that you want to call using the template
+ * Represents the common fields and attributes of an HTTP response.
+ * It contains also the original [Request] which was made to get this [Response].
  */
-fun http(baseUrl: String = "") = Request(baseUrl = baseUrl)
+open class Response(
+    private val response: FetchResponse,
+    val request: Request
+) {
+    val ok: Boolean get() = response.ok
+
+    val status: Int get() = response.status.toInt()
+
+    val url: String get() = response.url
+
+    val statusText: String get() = response.statusText
+
+    /**
+     * returns the [Headers] from the given [Response]
+     */
+    val headers get() = response.headers
+
+    /**
+     * extracts the body as string from the given [Response]
+     */
+    suspend fun body() = response.text().await()
+
+    /**
+     * extracts the body as blob from the given [Response]
+     */
+    suspend fun blob() = response.blob().await()
+
+    /**
+     * extracts the body as arrayBuffer from the given [Response]
+     */
+    suspend fun arrayBuffer() = response.arrayBuffer().await()
+
+    /**
+     * extracts the body as formData from the given [Response]
+     */
+    suspend fun formData() = response.formData().await()
+
+    /**
+     * extracts the body as json from the given [Response]
+     */
+    suspend fun json() = response.json().await()
+
+}
+
+/**
+ * interface to do interceptions at http calls.
+ * It can modify each request and handle all responses in a specified way.
+ */
+interface Interceptor {
```

Comment:
Lost of Other frameworks call this "middleware". I would suggest adapting to the common wording though.